[클린코드 3기 이현정] 계산기 미션

jlee0505 opened the pull request:
코드 리뷰 요청 드립니다

jlee0505 committed:
initial commit

jlee0505 committed:
feat: calculator.js 기능 요구사항 1차 완료

jlee0505 committed:
cypress installed

jlee0505 committed:
cypress test code trial 1

jlee0505 committed:
test

jlee0505 committed:
refactor: utils and handlers seperated from index.js

JunilHwang commented on cypress/e2e/spec.cy.js:
```diff
@@ -0,0 +1,67 @@
+
+  const clickOperator = (operator) => {
+    cy.get(".operation").contains(operator).click();
+  };
```
재사용하기 👍

JunilHwang commented on cypress/e2e/spec.cy.js:
```diff
@@ -0,0 +1,67 @@
+describe("test calculator", () => {
+  const clickDigit = (num) => {
+    cy.get(".digit").contains(num.toString()).click();
+  };
```
보통 이럴 땐 `handleClickDigit` 처럼 표현한답니다!

JunilHwang commented on cypress/e2e/spec.cy.js:
```diff
@@ -0,0 +1,67 @@
+    clickDigit(5);
+    clickOperator("=");
+    cy.get("#total").should("have.text", "128");
+  });
```
`test + operation` 의 경우 `click four digits`에 의존적인 테스트네요!
독립적인 테스트로 만들어보면 어떨까요?

JunilHwang commented on cypress/e2e/spec.cy.js:
```diff
@@ -0,0 +1,67 @@
+    cy.get(".digit").contains("9").click();
+    cy.get(".modifiers").click();
+    cy.get("#total").should("have.text", "0");
+  });
```
여기 있는 모든 테스트가 앞선 내용과 동일합니다!

JunilHwang replied:
그리고 전체적으로 테스트 케이스가 한 개 밖에 없네요. [each](https://jestjs.io/blog/2018/05/29/jest-23-blazing-fast-delightful-testing#jest-each) 등의 기능을 이용해서 여러 테스트 케이스를 만들어보세요!

JunilHwang commented on src/js/utils/DOM.js:
```diff
@@ -0,0 +1,5 @@
+export const $ = (elementName) => document.querySelector(elementName);
+export const TOTAL = $("#total");
+export const DIGITS = $(".digits");
+export const OPERATIONS = $(".operations");
+export const ALLCLEAR = $(".modifiers");
```
이렇게 DOM만 따로 모아서 관리한다면 코드의 응집도가 떨어지기 때문에 코드의 양이 많아질 수록 관리하기가 어려워질 수 있답니다!

그래서 DOM에 관련된 것들을 파일로 분리하기 보단,
기능 단위로 분리를 해보면 어떨까요?

JunilHwang commented on src/js/utils/constants.js:
```diff
@@ -0,0 +1,15 @@
+export const ERROR_MESSAGES = {
+  INVALID_DIGIT_LENGTH: `숫자는 최대 ${MAX_DIGIT_LENGTH}자리까지만 입력 가능합니다!`,
+  INVALID_INPUT: "숫자를 먼저 입력해 주세요!",
+};
```
앞선 내용과 동일합니다.
모든 상수를 한 파일에 몰아 놓게 된다면, 상수가 많아질수록 관리하기가 어려워진답니다!

JunilHwang commented on src/js/handlers/validator.js:
```diff
@@ -0,0 +1,15 @@
+
+export const isValidInput = () => {
+  return true;
+};
```
이 함수는 어떤 의미가 있는걸까요?

JunilHwang commented on src/js/handlers/validator.js:
```diff
@@ -0,0 +1,15 @@
+  if (!operator) {
+    return displayInput.length < MAX_DIGIT_LENGTH;
+  }
+};
```
여기서 DOM에 직접 접근하고 있네요!
validator의 경우 DOM에 종속적이지 않게 만들어주면 어떨까요?
가령, 계산기 미션을 브라우저 환경이 아니라 node.js 환경에서 만든다고 가정 했을 때, 이 validator에 있는 함수들을 재활용할 수 있을까요?

JunilHwang commented on src/js/handlers/handleCalClicked.js:
```diff
@@ -0,0 +1,52 @@
+  } else {
+    alert(`${ERROR_MESSAGES.INVALID_DIGIT_LENGTH}`);
+    return;
+  }
```
```suggestion
  if (isValidLength()) {
    TOTAL.textContent = digit;
    return;
  }
 alert(ERROR_MESSAGES.INVALID_DIGIT_LENGTH);
```
이렇게 표현해도 되겠네요!

JunilHwang commented on src/js/handlers/handleCalClicked.js:
```diff
@@ -0,0 +1,52 @@
+  } else {
+    alert(`${ERROR_MESSAGES.INVALID_INPUT}`);
+    return;
+  }
```
앞선 내용과 동일합니다!

JunilHwang commented on src/js/handlers/handleCalClicked.js:
```diff
@@ -0,0 +1,52 @@
+    .find((i) => OPERATORS.includes(i));
+  const nums = TOTAL.textContent.split(operator);
+  const num1 = nums[0];
+  const num2 = nums[1];
```
```suggestion
  const [ num1, num2 ] = TOTAL.textContent.split(operator);
```
이렇게 표현할 수 있답니다!

JunilHwang commented on src/js/handlers/handleCalClicked.js:
```diff
@@ -0,0 +1,52 @@
+    TOTAL.textContent = INITIAL_VALUE;
+    return;
+  }
+};
```
하나의 함수가 여러 가지 일을 하고 있네요!
내부에 있는 기능들을 별도의 함수로 만들어서 관리해주면 어떨까요?

JunilHwang commented on cypress/e2e/spec.cy.js:
```diff
@@ -0,0 +1,67 @@
+    cy.get(".digit").contains("9").click();
+    cy.get(".modifiers").click();
+    cy.get("#total").should("have.text", "0");
+  });
```
전체적으로 실패하는 테스트가 없네요.
실패에 대해 처리하는 테스트도 작성해주면 좋을 것 같아요!

JunilHwang reviewed (changes requested):
안녕하세요 현정님! 계산기 미션을 함께하게 된 리뷰어 황준일입니다 🙇‍♂️
계산기 미션 빠르게 잘 구현해주셨네요 👍

전체적으로 피드백을 남기자면

- 테스트
  - `it.each`를 이용해서 다양한 케이스에 대한 테스트를 만들어 보면 어떨까요?
  - 각각의 테스트가 다른 테스트의 결과와 상관 없이 독립적일 수 있도록 만들어주면 좋을 것 같아요!
  - 실패(오류)에 대한 테스트가 존재하지 않습니다. 이에 대한 테스트 작성도 필요하답니다!
    - 정확히는, 원래 테스트 코드를 작성할 때는 실패에 대한 코드 우선으로 작성해야 합니다 ㅎㅎ
- DOM, constants 등
  - 비슷한 성격의 코드를 한 파일에 몰아놓았는데, 만약에 사용해야 하는 DOM의 갯수가 100개, 상수도 100개 정도 라고 생각해보면(기능이 지금보다 많다면) 지금 처럼 구조를 가져갔을 때 관리하기가 힘들지 않을까요? 반대로 지금 방식이 관리하기 편하다고 생각했다면 어떤 이유 때문일까요?
- validator
  - validator 코드가 DOM에 종속적입니다. 만약에 브라우저가 아닌 환경(네이티브앱, node.js 등)에서 계산기 미션을 구현한다고 가정해본다면 이 validator를 활용할 수 있을까요?
  - 즉, 브라우저에 종속적인 코드와 브라우저 외부에서도 활용할 수 있는 코드를 분리해서 관리한다면 어떨까요?

핵심 적인 내용은 이정도입니다. 확인 후 다시 피드백 요청 부탁드려요!
감사합니다 😁🙇‍♂️

jlee0505 committed:
refactor

jlee0505 committed:
필요없는 else 문 제거

jlee0505 committed:
refactor

JunilHwang commented on cypress/e2e/spec.cy.js:
```diff
@@ -0,0 +1,70 @@
+  };
+
+  before(() => {
+    cy.visit("http://127.0.0.1:5500/");
```
별도로 서버를 실행시켜야 이 부분을 통과할 수 있네요!
`./index.html` 로 변경한다면 서버 실행 없이도 가능하답니다 ㅎㅎ

JunilHwang commented on cypress/e2e/spec.cy.js:
```diff
@@ -0,0 +1,70 @@
+    cy.get(".operation").contains(operator).click();
+  };
+
+  before(() => {
+    cy.visit("http://127.0.0.1:5500/");
+  });
```
그리고 다른 테스트가 통과하지 못하는 이유는,
첫번째 테스트가 통과된 다음에
UI가 초기화되지 않기 때문입니다 ㅎㅎ


```suggestion
  beforeEach(() => {
    cy.visit("./index.html");
  });
```
그래서 이렇게 `beforeEach`로 변경하면 된답니다!
before는 전체 테스트가 실행되기 전에 한 번 실행하는 함수고
beforeEach는 각각의 테스트가 실행되기 전에 실행해주는 함수입니다.

JunilHwang commented on src/js/handlers/handleCalClicked.js:
```diff
@@ -0,0 +1,58 @@
+import { operation } from "../utils/operation.js";
+
+const displayDigit = (digit) => {
+  if (!isValidDigitLength(TOTAL)) {
+    alert(`${ERROR_MESSAGES.INVALID_DIGIT_LENGTH}`);
+    return;
+  }
```
```suggestion
  if (!isValidDigitLength(TOTAL)) {
    alert(ERROR_MESSAGES.INVALID_DIGIT_LENGTH);
    return;
  }
```
이렇게 표현해도 되겠네요!

JunilHwang commented on src/js/handlers/handleCalClicked.js:
```diff
@@ -0,0 +1,58 @@
+  ERROR_MESSAGES,
+  OPERATORS,
+} from "../utils/constants.js";
+import { TOTAL, DIGITS, OPERATIONS, ALLCLEAR } from "../utils/DOM.js";
```
DOM이 utils에 해당하는것 같진 않네요 ㅎㅎ
utils이 정확히 어떤 개념일까요?

JunilHwang commented on src/js/handlers/handleCalClicked.js:
```diff
@@ -0,0 +1,58 @@
+import {
+  INITIAL_VALUE,
+  ERROR_MESSAGES,
+  OPERATORS,
+} from "../utils/constants.js";
```
constants도 앞선 내용과 동일합니다!

JunilHwang commented on src/js/handlers/handleCalClicked.js:
```diff
@@ -0,0 +1,58 @@
+} from "../utils/constants.js";
+import { TOTAL, DIGITS, OPERATIONS, ALLCLEAR } from "../utils/DOM.js";
+import { isAbleOperator, isValidDigitLength } from "./validator.js";
+import { operation } from "../utils/operation.js";
```
operation은 더더욱 utils이 아니랍니다!

JunilHwang commented on src/js/handlers/handleCalClicked.js:
```diff
@@ -0,0 +1,58 @@
+};
+
+// 하나의 함수가 여러 일을 하고 있다. -> 분리
+export const handleCalClicked = ({ target }) => {
+  if (target.classList.contains("digit")) {
+    displayDigit(target.textContent);
+    return;
+  }
+  if (target.classList.contains("operation")) {
+    displayOperation(target.textContent);
+    return;
+  }
+  if (target.classList.contains("modifier")) {
+    TOTAL.textContent = INITIAL_VALUE;
+    return;
+  }
+};
```
```js
addEvent(target, '.digit', displayDigit);
addEvent(target, '.operation', displayDigit);
addEvent(target, '.modifier', displayDigit);
```
이런식으로 사용할 수 있게 함수를 하나 정의해주면 어떨까요?

JunilHwang commented on src/js/handlers/handleCalClicked.js:
```diff
@@ -0,0 +1,58 @@
+    return;
+  }
+  alert(`${ERROR_MESSAGES.INVALID_INPUT}`);
+  return;
```
이 return은 없어도 무방하겠네요!

JunilHwang commented on src/js/handlers/handleCalClicked.js:
```diff
@@ -0,0 +1,58 @@
+};
+
+const displayOperation = (operator) => {
+  if (operator === "=") {
```
`=`는 상수로 만들어서 관리해주면 어떨까요?

JunilHwang commented on src/js/handlers/handleCalClicked.js:
```diff
@@ -0,0 +1,58 @@
+    alert(`${ERROR_MESSAGES.INVALID_DIGIT_LENGTH}`);
+    return;
+  }
+  if (TOTAL.textContent === INITIAL_VALUE) {
+    TOTAL.textContent = digit;
+  } else {
+    TOTAL.textContent += digit;
+  }
```
```suggestion
  TOTAL.textContent = Number(TOTAL.textContent + digit);
```
이렇게 표현해주면 어떨까요?
![image](https://user-images.githubusercontent.com/18749057/177332004-d9322277-0012-4b33-a72f-110736871a10.png)

JunilHwang commented on src/js/handlers/validator.js:
```diff
@@ -0,0 +1,26 @@
+import { TOTAL } from "../utils/DOM.js";
+
+// 브라우저에서 독립적으로 만들어주기
+export const isValidDigitLength = (e) => {
+  const displayInput = e.textContent;
+  const operator = displayInput.split("").find((i) => OPERATORS.includes(i));
+
+  if (!operator) {
+    return displayInput.length < MAX_DIGIT_LENGTH;
+  }
+
+  return displayInput.split(operator)?.pop().length < MAX_DIGIT_LENGTH;
+};
```
event 인자를 받아오고 있기 때문에 여전히 브라우저에 종속적인 코드랍니다!

JunilHwang replied:
validator가 하는 일이 총 3가지입니다.
- operator를 추출한다.
- 첫 번째 입력값을 검증한다.
- 세 번째 입력값을 검증한다.

이런 기능들을 더 쪼개서 함수로 만들어주면 어떨까요?
하나의 함수가 많은 일을 하게 되면 다른 사람이 코드를 이해하기가 쉽지 않답니다!

JunilHwang commented on src/js/handlers/validator.js:
```diff
@@ -0,0 +1,26 @@
+
+export const isAbleOperator = (e) => {
+  const displayInput = e.textContent;
+  if (displayInput === INITIAL_VALUE) {
+    return false;
+  }
+  return !Number.isNaN(typeof displayInput[displayInput.length - 1]);
```
```suggestion
  return (
    displayInput !== INITIAL_VALUE &&
    !Number.isNaN(typeof displayInput[displayInput.length - 1])
  );
```
이렇게 표현할 수 있겠네요!

그런데 지금 전체적으로 사용자의 입력값을 하나의 문자열로 만들고,
이 문자열을 검증하려는 시도가 보입니다.
문자열이 아니라 `자료구조`로 만들어서 관리해주면 어떨까요?

JunilHwang commented on src/js/utils/constants.js:
```diff
@@ -0,0 +1,15 @@
+export const MAX_DIGIT_LENGTH = 3;
+export const INITIAL_VALUE = "0";
+
+export const ADD = "+";
+export const SUB = "-";
+export const MUL = "X";
+export const DIV = "/";
+export const EQU = "=";
```
변수 이름에 줄여쓰기는 이양해주세요!

JunilHwang reviewed (approved):
안녕하세요 현정님!
리팩토링 하느라 고생하셨습니다 😁💪👍
다만.. 아직 고민해야할 부분이 무척 많이 있네요 ㅎㅎ
나중에 시간 되실 때 고쳐보시면 좋을 것 같습니다.
계산기 미션은 온보딩을 위한 미션이라서 이정도로 마무리 하도록 하겠습니다.
테스트코드가 정상적으로 실행되지 않는 부분도 리뷰에 언급했으니 확인 부탁드려요!
로또미션 진행하실 때 참고하시면 될 것 같습니다.

짧은 기간이었으나 부디 조금이나마 도움이 되었길 바랍니다 😁
감사합니다!

나머지 미션도 무사 완주 기원합니다 🙏

jlee0505 commented:
아앗...역시 그렇죠ㅠㅠ 이번 미션부터는 좀 더 부지런히 시작해보아야겠습니다. 친절+빠른+꼼꼼한 리뷰 감사드립니다!!!

On Tue, Jul 5, 2022, 10:02 PM 황준일 ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> 안녕하세요 현정님!
> 리팩토링 하느라 고생하셨습니다 😁💪👍
> 다만.. 아직 고민해야할 부분이 무척 많이 있네요 ㅎㅎ
> 나중에 시간 되실 때 고쳐보시면 좋을 것 같습니다.
> 계산기 미션은 온보딩을 위한 미션이라서 이정도로 마무리 하도록 하겠습니다.
> 테스트코드가 정상적으로 실행되지 않는 부분도 리뷰에 언급했으니 확인 부탁드려요!
> 로또미션 진행하실 때 참고하시면 될 것 같습니다.
>
> 짧은 기간이었으나 부디 조금이나마 도움이 되었길 바랍니다 😁
> 감사합니다!
>
> 나머지 미션도 무사 완주 기원합니다 🙏
> ------------------------------
>
> In cypress/e2e/spec.cy.js
> <https://github.com/next-step/js-calculator/pull/122#discussion_r913751327>
> :
>
> > @@ -0,0 +1,70 @@
>
> +Cypress.on("uncaught:exception", (err, runnable) => {
>
> +  return false;
>
> +});
>
> +
>
> +describe("test calculator", () => {
>
> +  const handleClickDigit = (num) => {
>
> +    cy.get(".digit").contains(num.toString()).click();
>
> +  };
>
> +
>
> +  const handleClickOperator = (operator) => {
>
> +    cy.get(".operation").contains(operator).click();
>
> +  };
>
> +
>
> +  before(() => {
>
> +    cy.visit("http://127.0.0.1:5500/");
>
>
> 별도로 서버를 실행시켜야 이 부분을 통과할 수 있네요!
> ./index.html 로 변경한다면 서버 실행 없이도 가능하답니다 ㅎㅎ
> ------------------------------
>
> In cypress/e2e/spec.cy.js
> <https://github.com/next-step/js-calculator/pull/122#discussion_r913756191>
> :
>
> > +  before(() => {
>
> +    cy.visit("http://127.0.0.1:5500/");
>
> +  });
>
>
> 그리고 다른 테스트가 통과하지 못하는 이유는,
> 첫번째 테스트가 통과된 다음에
> UI가 초기화되지 않기 때문입니다 ㅎㅎ
> ⬇️ Suggested change
>
> -  before(() => {
>
> -    cy.visit("http://127.0.0.1:5500/");
>
> -  });
>
> +  beforeEach(() => {
>
> +    cy.visit("./index.html");
>
> +  });
>
>
> 그래서 이렇게 beforeEach로 변경하면 된답니다!
> before는 전체 테스트가 실행되기 전에 한 번 실행하는 함수고
> beforeEach는 각각의 테스트가 실행되기 전에 실행해주는 함수입니다.
> ------------------------------
>
> In src/js/handlers/handleCalClicked.js
> <https://github.com/next-step/js-calculator/pull/122#discussion_r913756671>
> :
>
> > +  if (!isValidDigitLength(TOTAL)) {
>
> +    alert(`${ERROR_MESSAGES.INVALID_DIGIT_LENGTH}`);
>
> +    return;
>
> +  }
>
>
> ⬇️ Suggested change
>
> -  if (!isValidDigitLength(TOTAL)) {
>
> -    alert(`${ERROR_MESSAGES.INVALID_DIGIT_LENGTH}`);
>
> -    return;
>
> -  }
>
> +  if (!isValidDigitLength(TOTAL)) {
>
> +    alert(ERROR_MESSAGES.INVALID_DIGIT_LENGTH);
>
> +    return;
>
> +  }
>
>
> 이렇게 표현해도 되겠네요!
> ------------------------------
>
> In src/js/handlers/handleCalClicked.js
> <https://github.com/next-step/js-calculator/pull/122#discussion_r913756959>
> :
>
> > @@ -0,0 +1,58 @@
>
> +import {
>
> +  INITIAL_VALUE,
>
> +  ERROR_MESSAGES,
>
> +  OPERATORS,
>
> +} from "../utils/constants.js";
>
> +import { TOTAL, DIGITS, OPERATIONS, ALLCLEAR } from "../utils/DOM.js";
>
>
> DOM이 utils에 해당하는것 같진 않네요 ㅎㅎ
> utils이 정확히 어떤 개념일까요?
> ------------------------------
>
> In src/js/handlers/handleCalClicked.js
> <https://github.com/next-step/js-calculator/pull/122#discussion_r913757142>
> :
>
> > +import {
>
> +  INITIAL_VALUE,
>
> +  ERROR_MESSAGES,
>
> +  OPERATORS,
>
> +} from "../utils/constants.js";
>
>
> constants도 앞선 내용과 동일합니다!
> ------------------------------
>
> In src/js/handlers/handleCalClicked.js
> <https://github.com/next-step/js-calculator/pull/122#discussion_r913757262>
> :
>
> > @@ -0,0 +1,58 @@
>
> +import {
>
> +  INITIAL_VALUE,
>
> +  ERROR_MESSAGES,
>
> +  OPERATORS,
>
> +} from "../utils/constants.js";
>
> +import { TOTAL, DIGITS, OPERATIONS, ALLCLEAR } from "../utils/DOM.js";
>
> +import { isAbleOperator, isValidDigitLength } from "./validator.js";
>
> +import { operation } from "../utils/operation.js";
>
>
> operation은 더더욱 utils이 아니랍니다!
> ------------------------------
>
> In src/js/handlers/handleCalClicked.js
> <https://github.com/next-step/js-calculator/pull/122#discussion_r913759527>
> :
>
> > +export const handleCalClicked = ({ target }) => {
>
> +  if (target.classList.contains("digit")) {
>
> +    displayDigit(target.textContent);
>
> +    return;
>
> +  }
>
> +  if (target.classList.contains("operation")) {
>
> +    displayOperation(target.textContent);
>
> +    return;
>
> +  }
>
> +  if (target.classList.contains("modifier")) {
>
> +    TOTAL.textContent = INITIAL_VALUE;
>
> +    return;
>
> +  }
>
> +};
>
>
> addEvent(target, '.digit', displayDigit);
> addEvent(target, '.operation', displayDigit);
> addEvent(target, '.modifier', displayDigit);
>
> 이런식으로 사용할 수 있게 함수를 하나 정의해주면 어떨까요?
> ------------------------------
>
> In src/js/handlers/handleCalClicked.js
> <https://github.com/next-step/js-calculator/pull/122#discussion_r913760978>
> :
>
> > +    TOTAL.textContent += digit;
>
> +  }
>
> +};
>
> +
>
> +const displayOperation = (operator) => {
>
> +  if (operator === "=") {
>
> +    displayTotal();
>
> +    return;
>
> +  }
>
> +
>
> +  if (isAbleOperator(TOTAL)) {
>
> +    TOTAL.innerText += operator;
>
> +    return;
>
> +  }
>
> +  alert(`${ERROR_MESSAGES.INVALID_INPUT}`);
>
> +  return;
>
>
> 이 return은 없어도 무방하겠네요!
> ------------------------------
>
> In src/js/handlers/handleCalClicked.js
> <https://github.com/next-step/js-calculator/pull/122#discussion_r913761117>
> :
>
> > +import { operation } from "../utils/operation.js";
>
> +
>
> +const displayDigit = (digit) => {
>
> +  if (!isValidDigitLength(TOTAL)) {
>
> +    alert(`${ERROR_MESSAGES.INVALID_DIGIT_LENGTH}`);
>
> +    return;
>
> +  }
>
> +  if (TOTAL.textContent === INITIAL_VALUE) {
>
> +    TOTAL.textContent = digit;
>
> +  } else {
>
> +    TOTAL.textContent += digit;
>
> +  }
>
> +};
>
> +
>
> +const displayOperation = (operator) => {
>
> +  if (operator === "=") {
>
>
> =는 상수로 만들어서 관리해주면 어떨까요?
> ------------------------------
>
> In src/js/handlers/handleCalClicked.js
> <https://github.com/next-step/js-calculator/pull/122#discussion_r913762475>
> :
>
> > +  if (TOTAL.textContent === INITIAL_VALUE) {
>
> +    TOTAL.textContent = digit;
>
> +  } else {
>
> +    TOTAL.textContent += digit;
>
> +  }
>
>
> ⬇️ Suggested change
>
> -  if (TOTAL.textContent === INITIAL_VALUE) {
>
> -    TOTAL.textContent = digit;
>
> -  } else {
>
> -    TOTAL.textContent += digit;
>
> -  }
>
> +  TOTAL.textContent = Number(TOTAL.textContent + digit);
>
>
> 이렇게 표현해주면 어떨까요?
> [image: image]
> <https://user-images.githubusercontent.com/18749057/177332004-d9322277-0012-4b33-a72f-110736871a10.png>
> ------------------------------
>
> In src/js/handlers/validator.js
> <https://github.com/next-step/js-calculator/pull/122#discussion_r913764960>
> :
>
> > +export const isValidDigitLength = (e) => {
>
> +  const displayInput = e.textContent;
>
> +  const operator = displayInput.split("").find((i) => OPERATORS.includes(i));
>
> +
>
> +  if (!operator) {
>
> +    return displayInput.length < MAX_DIGIT_LENGTH;
>
> +  }
>
> +
>
> +  return displayInput.split(operator)?.pop().length < MAX_DIGIT_LENGTH;
>
> +};
>
>
> event 인자를 받아오고 있기 때문에 여전히 브라우저에 종속적인 코드랍니다!
> ------------------------------
>
> In src/js/handlers/validator.js
> <https://github.com/next-step/js-calculator/pull/122#discussion_r913767177>
> :
>
> > +export const isValidDigitLength = (e) => {
>
> +  const displayInput = e.textContent;
>
> +  const operator = displayInput.split("").find((i) => OPERATORS.includes(i));
>
> +
>
> +  if (!operator) {
>
> +    return displayInput.length < MAX_DIGIT_LENGTH;
>
> +  }
>
> +
>
> +  return displayInput.split(operator)?.pop().length < MAX_DIGIT_LENGTH;
>
> +};
>
>
> validator가 하는 일이 총 3가지입니다.
>
>    - operator를 추출한다.
>    - 첫 번째 입력값을 검증한다.
>    - 세 번째 입력값을 검증한다.
>
> 이런 기능들을 더 쪼개서 함수로 만들어주면 어떨까요?
> 하나의 함수가 많은 일을 하게 되면 다른 사람이 코드를 이해하기가 쉽지 않답니다!
> ------------------------------
>
> In src/js/handlers/validator.js
> <https://github.com/next-step/js-calculator/pull/122#discussion_r913768378>
> :
>
> > +  if (displayInput === INITIAL_VALUE) {
>
> +    return false;
>
> +  }
>
> +  return !Number.isNaN(typeof displayInput[displayInput.length - 1]);
>
>
> ⬇️ Suggested change
>
> -  if (displayInput === INITIAL_VALUE) {
>
> -    return false;
>
> -  }
>
> -  return !Number.isNaN(typeof displayInput[displayInput.length - 1]);
>
> +  return (
>
> +    displayInput !== INITIAL_VALUE &&
>
> +    !Number.isNaN(typeof displayInput[displayInput.length - 1])
>
> +  );
>
>
> 이렇게 표현할 수 있겠네요!
>
> 그런데 지금 전체적으로 사용자의 입력값을 하나의 문자열로 만들고,
> 이 문자열을 검증하려는 시도가 보입니다.
> 문자열이 아니라 자료구조로 만들어서 관리해주면 어떨까요?
> ------------------------------
>
> In src/js/utils/constants.js
> <https://github.com/next-step/js-calculator/pull/122#discussion_r913769044>
> :
>
> > +export const ADD = "+";
>
> +export const SUB = "-";
>
> +export const MUL = "X";
>
> +export const DIV = "/";
>
> +export const EQU = "=";
>
>
> 변수 이름에 줄여쓰기는 이양해주세요!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/next-step/js-calculator/pull/122#pullrequestreview-1028566857>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM7W2LACUSZG2UJN4J67ZFDVSQW6BANCNFSM52PC547Q>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>
